### PR TITLE
vCenter push capability

### PR DESCRIPTION
**Pull request by thebsdbox**
Re-factored the `run` code to create the `push` functionality.

Signed-off-by: Dan Finneran <daniel.finneran@gmail.com>

**- What I did**
Added the capability to `push` Linuxkit `.iso` images to VMware vCenter. It uses a either a `-folder=<NAME>` or will take the name of the image and create that folder to host the uploaded image.

**- How I did it**
Modified `run_vcenter.go` to allow `push_vcenter.go` to log in to vCenter and upload images without creating a full VM. 

**- How to verify it**
`Push` functionality
```
./bin/linuxkit push vcenter -datastore=vSphereNFS -hostname=esxi01.fnnrn.me -folder=sharedLinuxKit2  ./vcenter.iso 
Uploading LinuxKit file [./vcenter.iso]
```
`run` functionality (ensuring changes haven't broken existing functionality.
```
./bin/linuxkit run vcenter -datastore=vSphereNFS -network=HomeNetwork -hostname=esxi01.fnnrn.me -powerOn -persistentSize=1024 ./vcenter.iso 
Creating new LinuxKit Virtual Machine
Uploading LinuxKit file [./vcenter.iso]
Adding ISO to the Virtual Machine
Adding a persistent disk to the Virtual Machine
Adding VM Networking
Powering on LinuxKit VM

```

**- Description for the changelog**

Added the functionality to `push` linuxkit images to VMware vCenter.

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: src/cmd/linuxkit/push_vcenter.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	flags := flag.NewFlagSet("vCenter", flag.ExitOnError)
+	invoked := filepath.Base(os.Args[0])
+
+	newVM.vCenterURL = flags.String("url", os.Getenv("VCURL"), "URL in the format of https://username:password@host/sdk")
+	newVM.dsName = flags.String("datastore", os.Getenv("VMDATASTORE"), "The Name of the DataStore to host the image")
+	newVM.networkName = flags.String("network", os.Getenv("VMNETWORK"), "The VMware vSwitch the image will use")
+	newVM.vSphereHost = flags.String("hostname", os.Getenv("VMHOST"), "The Server that will host the image")
+	newVM.path = flags.String("path", "", "Path to a specific image")
+
+	newVM.vmName = flags.String("folder", "", "A Folder on the datastore to push the image too")
```
Comment (justincormack) [inline src/cmd/linuxkit/push_vcenter.go]: its confusing calling it "folder" and "vmName" - maybe call both folder?

File: src/cmd/linuxkit/push_vcenter.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+		fmt.Printf("Please specify the path to the image to push\n")
+		flags.Usage()
+		os.Exit(1)
+	}
+	*newVM.path = remArgs[0]
+
+	// Ensure an iso has been passed to the vCenter push Command
+	if strings.HasSuffix(*newVM.path, ".iso") {
+		// Used as a folder to hold the images that are pushed
+		if *newVM.vmName == "" {
```
Comment (justincormack) [inline src/cmd/linuxkit/push_vcenter.go]: there is no need for this to be in this condition, it can be unconditional below.

File: src/cmd/linuxkit/push_vcenter.go (hunk 1 above)
Comment (thebsdbox) [inline src/cmd/linuxkit/push_vcenter.go]: Opted for vmFolder, as it makes sense in both use cases. With the `run` the vmFolder is both the name of the VM and the name of the folder that will be created to house the various VM files. With `push` the vmFolder is simply a folder created on a datastore that a will hold the VM `iso` file. Does that sound OK?

File: src/cmd/linuxkit/push_vcenter.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	var newVM vmConfig
+
+	flags := flag.NewFlagSet("vCenter", flag.ExitOnError)
+	invoked := filepath.Base(os.Args[0])
+
+	newVM.vCenterURL = flags.String("url", os.Getenv("VCURL"), "URL in the format of https://username:password@host/sdk")
+	newVM.dsName = flags.String("datastore", os.Getenv("VMDATASTORE"), "The Name of the DataStore to host the image")
+	newVM.networkName = flags.String("network", os.Getenv("VMNETWORK"), "The VMware vSwitch the image will use")
+	newVM.vSphereHost = flags.String("hostname", os.Getenv("VMHOST"), "The Server that will host the image")
+	newVM.path = flags.String("path", "", "Path to a specific image")
```
Comment (rneugeba) [inline src/cmd/linuxkit/push_vcenter.go]: Are these standard environment variables or made up?  They seem to be inconsistently named and I would prefer if they had a common prefix...

File: src/cmd/linuxkit/push_vcenter.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+func pushVCenter(args []string) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	var newVM vmConfig
+
+	flags := flag.NewFlagSet("vCenter", flag.ExitOnError)
+	invoked := filepath.Base(os.Args[0])
+
+	newVM.vCenterURL = flags.String("url", os.Getenv("VCURL"), "URL in the format of https://username:password@host/sdk")
```
Comment (rneugeba) [inline src/cmd/linuxkit/push_vcenter.go]: Is this the URL to the vcenter host? Id so, mention it in the help message

File: src/cmd/linuxkit/push_vcenter.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	newVM.dsName = flags.String("datastore", os.Getenv("VMDATASTORE"), "The Name of the DataStore to host the image")
```
Comment (rneugeba) [inline src/cmd/linuxkit/push_vcenter.go]: nit: Name -> name

File: src/cmd/linuxkit/push_vcenter.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+	newVM.networkName = flags.String("network", os.Getenv("VMNETWORK"), "The VMware vSwitch the image will use")
```
Comment (rneugeba) [inline src/cmd/linuxkit/push_vcenter.go]: Is this the name of the network the VM will be placed on or the name of a vSwitch as the help message implies?

File: src/cmd/linuxkit/push_vcenter.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+	newVM.vSphereHost = flags.String("hostname", os.Getenv("VMHOST"), "The Server that will host the image")
```
Comment (rneugeba) [inline src/cmd/linuxkit/push_vcenter.go]: nit: Server -> server

File: src/cmd/linuxkit/push_vcenter.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+	newVM.path = flags.String("path", "", "Path to a specific image")
+
+	newVM.vmFolder = flags.String("folder", "", "A Folder on the datastore to push the image too")
+
+	flags.Usage = func() {
+		fmt.Printf("USAGE: %s push vcenter [options] [name]\n\n", invoked)
```
Comment (rneugeba) [inline src/cmd/linuxkit/push_vcenter.go]: is `name` optional as the brackets imply? To be consistent with the qemu runner this should probably be called `path`

File: src/cmd/linuxkit/push_vcenter.go [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+	if len(remArgs) == 0 {
+		fmt.Printf("Please specify the path to the image to push\n")
+		flags.Usage()
+		os.Exit(1)
+	}
+	*newVM.path = remArgs[0]
+
+	// Ensure an iso has been passed to the vCenter push Command
+	if !strings.HasSuffix(*newVM.path, ".iso") {
+		log.Fatalln("Ensure that an \".iso\" file is used as part of the path")
```
Comment (rneugeba) [inline src/cmd/linuxkit/push_vcenter.go]: maybe rephrase: `Please pass an ".iso" file as the path` or somesuch.

File: src/cmd/linuxkit/push_vcenter.go [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+	flags.Usage = func() {
+		fmt.Printf("USAGE: %s push vcenter [options] path \n\n", invoked)
+		fmt.Printf("'path' specifies the full path of an image that will be pushed\n")
+		fmt.Printf("Options:\n\n")
+		flags.PrintDefaults()
+	}
+
+	if err := flags.Parse(args); err != nil {
+		log.Fatalln("Unable to parse args")
```
Comment (riyazdf) [inline src/cmd/linuxkit/push_vcenter.go]: should we include the error message?  `log.Fatalf("Unable to parse args: %v", err)`?

File: src/cmd/linuxkit/push_vcenter.go [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+	flags := flag.NewFlagSet("vCenter", flag.ExitOnError)
+	invoked := filepath.Base(os.Args[0])
+
+	newVM.vCenterURL = flags.String("url", os.Getenv("VCURL"), "URL of VMware vCenter in the format of https://username:password@VCaddress/sdk")
+	newVM.dsName = flags.String("datastore", os.Getenv("VCDATASTORE"), "The name of the DataStore to host the image")
+	newVM.networkName = flags.String("network", os.Getenv("VCNETWORK"), "The network label the VM will use")
+	newVM.vSphereHost = flags.String("hostname", os.Getenv("VCHOST"), "The server that will host the image")
+	newVM.path = flags.String("path", "", "Path to a specific image")
+
+	newVM.vmFolder = flags.String("folder", "", "A Folder on the datastore to push the image too")
```
Comment (riyazdf) [inline src/cmd/linuxkit/push_vcenter.go]: nit: can we lowercase to `folder`?

File: src/cmd/linuxkit/run_vcenter.go [hunk 12]
```diff
@@ -43,21 +43,21 @@ func runVcenter(args []string) {
...
-	newVM.dsName = flags.String("datastore", os.Getenv("VMDATASTORE"), "The Name of the DataStore to host the VM")
-	newVM.networkName = flags.String("network", os.Getenv("VMNETWORK"), "The VMware vSwitch the VM will use")
-	newVM.vSphereHost = flags.String("hostname", os.Getenv("VMHOST"), "The Server that will run the VM")
+	newVM.vCenterURL = flags.String("url", os.Getenv("VCURL"), "URL of VMware vCenter in the format of https://username:password@VCaddress/sdk")
+	newVM.dsName = flags.String("datastore", os.Getenv("VCDATASTORE"), "The name of the DataStore to host the VM")
+	newVM.networkName = flags.String("network", os.Getenv("VCNETWORK"), "The network label the VM will use")
+	newVM.vSphereHost = flags.String("hostname", os.Getenv("VCHOST"), "The server that will run the VM")
 
-	newVM.vmName = flags.String("vmname", "", "Specify a name for virtual Machine")
+	newVM.vmFolder = flags.String("vmfolder", "", "Specify a name/folder for the virtual Machine to reside in")
```
Comment (riyazdf) [inline src/cmd/linuxkit/run_vcenter.go]: nit: can we lowercase to `machine`?

File: src/cmd/linuxkit/run_vcenter.go [hunk 13]
```diff
@@ -43,21 +43,21 @@ func runVcenter(args []string) {
...
 	invoked := filepath.Base(os.Args[0])
 
-	newVM.vCenterURL = flags.String("url", os.Getenv("VCURL"), "URL in the format of https://username:password@host/sdk")
-	newVM.dsName = flags.String("datastore", os.Getenv("VMDATASTORE"), "The Name of the DataStore to host the VM")
-	newVM.networkName = flags.String("network", os.Getenv("VMNETWORK"), "The VMware vSwitch the VM will use")
-	newVM.vSphereHost = flags.String("hostname", os.Getenv("VMHOST"), "The Server that will run the VM")
+	newVM.vCenterURL = flags.String("url", os.Getenv("VCURL"), "URL of VMware vCenter in the format of https://username:password@VCaddress/sdk")
+	newVM.dsName = flags.String("datastore", os.Getenv("VCDATASTORE"), "The name of the DataStore to host the VM")
+	newVM.networkName = flags.String("network", os.Getenv("VCNETWORK"), "The network label the VM will use")
+	newVM.vSphereHost = flags.String("hostname", os.Getenv("VCHOST"), "The server that will run the VM")
```
Comment (riyazdf) [inline src/cmd/linuxkit/run_vcenter.go]: what is the reason for changing to `VC*`?  Should we update the other flags like `vmfolder`?  I'm not too familiar with vCenter so I'm not sure what makes most sense here :)

File: src/cmd/linuxkit/run_vcenter.go [hunk 14]
```diff
@@ -191,6 +143,66 @@ func runVcenter(args []string) {
 	}
 }
 
+func vCenterConnect(ctx context.Context, newVM vmConfig) (*govmomi.Client, *object.Datastore, *object.DatacenterFolders, *object.HostSystem, object.NetworkReference, *object.ResourcePool) {
+
+	// Parse URL from string
+	u, err := url.Parse(*newVM.vCenterURL)
+	if err != nil {
+		log.Fatalf("URL can't be parsed, ensure it is https://username:password/<address>/sdk")
```
Comment (riyazdf) [inline src/cmd/linuxkit/run_vcenter.go]: might be helpful to include the err message here?  `log.Fatalf("URL can't be parsed, ensure it is https://username:password/<address>/sdk: %v", err)`

File: src/cmd/linuxkit/run_vcenter.go [hunk 15]
```diff
@@ -191,6 +143,66 @@ func runVcenter(args []string) {
...
+	}
+
+	// Connect and log in to ESX or vCenter
+	c, err := govmomi.NewClient(ctx, u, true)
+	if err != nil {
+		log.Fatalln("Error logging into vCenter, check address and credentials")
```
Comment (riyazdf) [inline src/cmd/linuxkit/run_vcenter.go]: same comment re: including error message: `log.Fatalf("Error logging into vCenter, check address and credentials: %v", err)`

File: src/cmd/linuxkit/run_vcenter.go [hunk 16]
```diff
@@ -191,6 +143,66 @@ func runVcenter(args []string) {
...
+	}
+
+	// Create a new finder that will discover the defaults and are looked for Networks/Datastores
+	f := find.NewFinder(c.Client, true)
+
+	// Find one and only datacenter, not sure how VMware linked mode will work
+	dc, err := f.DefaultDatacenter(ctx)
+	if err != nil {
+		log.Fatalln("No Datacenter instance could be found inside of vCenter")
```
Comment (riyazdf) [inline src/cmd/linuxkit/run_vcenter.go]: same comment re: including error message for this and subsequent errors

File: src/cmd/linuxkit/run_vcenter.go [hunk 17]
```diff
@@ -248,7 +260,7 @@ func addVMDK(ctx context.Context, vm *object.VirtualMachine, dss *object.Datasto
 		log.Fatalf("Unable to find SCSI device from VM configuration\n%v", err)
 	}
 	// The default is to have all persistent disks named linuxkit.vmdk
-	disk := devices.CreateDisk(controller, dss.Reference(), dss.Path(fmt.Sprintf("%s/%s", *newVM.vmName, "linuxkit.vmdk")))
+	disk := devices.CreateDisk(controller, dss.Reference(), dss.Path(fmt.Sprintf("%s/%s", *newVM.vmFolder, "linuxkit.vmdk")))
```
Comment (riyazdf) [inline src/cmd/linuxkit/run_vcenter.go]: should this become configurable?  Can be a separate PR, but wondering if we should open a tracking issue

File: src/cmd/linuxkit/run_vcenter.go (hunk 13 above)
Comment (thebsdbox) [inline src/cmd/linuxkit/run_vcenter.go]: The VC**** environment variables are there to make it a bit more clearer when repeatedly deploying VMs. The credentials and labels for vCenter internals will stay the same, so we can store them elsewhere (env vars in this case), the rest of the flags we can pass as usual as there is some need for uniqueness. Also @rneugeba, suggested some format for the env variables so I opted for VC<resource type>.

File: src/cmd/linuxkit/push_vcenter.go (hunk 10 above)
Comment (thebsdbox) [inline src/cmd/linuxkit/push_vcenter.go]: We should probably standardise this across all of the backends?

Comment (thebsdbox) [inline src/cmd/linuxkit/push_vcenter.go]: Actually, looking through the source code as we use `ExitOnError` and the `parse` does the following: 
```
		case ExitOnError:
			os.Exit(2)
```
The `log.fatalln(...)` will never be called.

File: src/cmd/linuxkit/run_vcenter.go (hunk 17 above)
Comment (thebsdbox) [inline src/cmd/linuxkit/run_vcenter.go]: It was to begin with, but then it became apparent that uploading  a`vmdk` from workstation to server required a lot more steps (image conversion from thin to fat, meant uploading GBs of zeros etc.) As there is a 1:1 mapping inside of VMware vCenter VMs I opted to keep it simple and in the event the `run` requires a persistent disk they just get a new disk inside of their VM folder to use. 

Happy to revisit this, it felt that uploading large existing disks or having to manipulate the disk controllers to support additional SCSI disks might be outside of what `run` should be doing :)

File: src/cmd/linuxkit/push_vcenter.go (hunk 10 above)
Comment (thebsdbox) [inline src/cmd/linuxkit/push_vcenter.go]: Does that look right @justincormack / @rneugeba / @riyazdf  ?

As we have this same snippet in a lot of places.

File: src/cmd/linuxkit/run_vcenter.go [hunk 18]
```diff
@@ -43,21 +43,21 @@ func runVcenter(args []string) {
...
 	newVM.persistent = flags.String("persistentSize", "", "Size in MB of persistent storage to allocate to the VM")
 	newVM.mem = flags.Int64("mem", 1024, "Size in MB of memory to allocate to the VM")
 	newVM.vCpus = flags.Int("cpus", 1, "Amount of vCPUs to allocate to the VM")
 	newVM.poweron = flags.Bool("powerOn", false, "Power On the new VM once it has been created")
 
 	flags.Usage = func() {
-		fmt.Printf("USAGE: %s run vcenter [options] [name]\n\n", invoked)
-		fmt.Printf("'name' specifies the full path of an image that will be ran\n")
+		fmt.Printf("USAGE: %s run vcenter [options] path\n\n", invoked)
+		fmt.Printf("'path' specifies the full path of an image that will be ran\n")
```
Comment (justincormack) [inline src/cmd/linuxkit/run_vcenter.go]: "that will be ran" reads weirdly, "to run" is better

File: src/cmd/linuxkit/push_vcenter.go (hunk 10 above)
Comment (thebsdbox) [inline src/cmd/linuxkit/push_vcenter.go]: OK, i'll leave it as is for now.

File: src/cmd/linuxkit/run_vcenter.go (hunk 18 above)
Comment (thebsdbox) [inline src/cmd/linuxkit/run_vcenter.go]: Updating, i've used "to run" below so changing it keeps the language consistent.

File: src/cmd/linuxkit/push_vcenter.go (hunk 10 above)
Comment (riyazdf) [inline src/cmd/linuxkit/push_vcenter.go]: got it - yeah we can probably standardize this with a custom error type but this is good for now

File: src/cmd/linuxkit/run_vcenter.go (hunk 13 above)
Comment (riyazdf) [inline src/cmd/linuxkit/run_vcenter.go]: :+1: SGTM

File: src/cmd/linuxkit/run_vcenter.go [hunk 19]
```diff
@@ -191,6 +143,66 @@ func runVcenter(args []string) {
...
+		log.Fatalf("Error logging into vCenter, check address and credentials %v", err)
+	}
+
+	// Create a new finder that will discover the defaults and are looked for Networks/Datastores
+	f := find.NewFinder(c.Client, true)
+
+	// Find one and only datacenter, not sure how VMware linked mode will work
+	dc, err := f.DefaultDatacenter(ctx)
+	if err != nil {
+		log.Fatalln("No Datacenter instance could be found inside of vCenter %v", err)
```
Comment (riyazdf) [inline src/cmd/linuxkit/run_vcenter.go]: nit: I think this should be `log.Fatalf`?

File: src/cmd/linuxkit/run_vcenter.go (hunk 17 above)
Comment (riyazdf) [inline src/cmd/linuxkit/run_vcenter.go]: ah got it.  Makes sense to keep it like this for now then, thank you for the explanation!

File: src/cmd/linuxkit/run_vcenter.go (hunk 19 above)
Comment (thebsdbox) [inline src/cmd/linuxkit/run_vcenter.go]: true :D fixed.
